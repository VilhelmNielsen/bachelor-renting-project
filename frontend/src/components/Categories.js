/** @jsx jsx */
import { useEffect } from 'react'
import { jsx, css } from '@emotion/core'
import { Link } from '@reach/router'
import gql from 'graphql-tag'
import { useQuery } from 'react-apollo-hooks'
import { siteMeta } from '../config'

import Layout from './Layout'

function Categories() {
  const {
    data: { categories },
    refetch,
  } = useQuery(QUERY_CATEGORIES, { suspend: true })

  useEffect(() => {
    document.title = 'Categories'
    return () => (document.title = siteMeta.title)
  }, [])

  useEffect(() => {
    refetch()
  }, [])

  return (
    <Layout>
      <section>
        <h1>Categories</h1>
        {categories &&
          categories.map(category => (
            <CategoryCard category={category} key={category.slug} />
          ))}
      </section>
    </Layout>
  )
}

const QUERY_CATEGORIES = gql`
  query categoriesPage {
    categories {
      id
      slug
      title
    }
  }
`

const cardStyles = {
  card: css`
    padding: 1rem;
  `,
}

function CategoryCard(props) {
  const { category } = props
  return (
    <div css={cardStyles.card}>
      <Link to={`/category/${category.slug}`}>{category.title}</Link>
    </div>
  )
}

export default Categories
export { QUERY_CATEGORIES }
