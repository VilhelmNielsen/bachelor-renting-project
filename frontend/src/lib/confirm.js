const confirm = window.confirm

export default confirm
