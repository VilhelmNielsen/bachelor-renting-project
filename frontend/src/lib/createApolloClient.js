import ApolloClient from 'apollo-boost'

function createApolloClient() {
  return new ApolloClient({
    uri: process.env.REACT_APP_GRAPHQL_ENDPOINT,
    credentials: 'include',
  })
}

export default createApolloClient
