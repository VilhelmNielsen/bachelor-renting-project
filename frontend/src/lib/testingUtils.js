import casual from 'casual'
import { ApolloClient } from 'apollo-client'
import { InMemoryCache } from 'apollo-cache-inmemory'
import { MockLink } from 'apollo-link-mock'

casual.seed(123)

export const fakeUser = overrides => ({
  __typename: 'User',
  id: 'abc123',
  name: casual.full_name,
  email: casual.email,
  items: [],
  bookings: [],
  pendingBookings: [],
  ...overrides,
})

export const fakeItem = overrides => ({
  __typename: 'Item',
  id: 'xyz789',
  title: casual.title,
  description: casual.description,
  price: 50000,
  averageRating: null,
  image: {
    full: 'dog_full.jpg',
    preview: 'dog_preview.jpg',
  },
  reviews: [],
  category: [fakeCategory()],
  ...overrides,
})

export const fakeCategory = overrides => ({
  __typename: 'Category',
  id: 'cat345',
  title: casual.title,
  slug: casual.word,
  description: casual.description,
  items: [],
  ...overrides,
})

export function createClient(mocks) {
  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new MockLink(mocks),
  })
}
