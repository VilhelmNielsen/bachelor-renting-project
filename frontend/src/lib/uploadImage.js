async function uploadImage(event) {
  console.log('Uploading image...')
  const {
    files: [file],
  } = event.target

  const data = new FormData()
  data.append('file', file)
  data.append('upload_preset', 'bachelor-project')

  const res = await fetch(
    'https://api.cloudinary.com/v1_1/vilhelmnielsen/image/upload',
    { method: 'POST', body: data }
  )

  console.log('Upload complete')

  return await res.json()
}

export default uploadImage
