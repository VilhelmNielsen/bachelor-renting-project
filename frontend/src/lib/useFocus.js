import { useState, useRef } from 'react'

function useFocus() {
  const [hasFocus, setHasFocus] = useState(false)
  const timer = useRef()

  const onFocus = event => {
    clearTimeout(timer.current)
    setHasFocus(true)
  }
  const onBlur = event => {
    timer.current = setTimeout(() => setHasFocus(false))
  }

  return [hasFocus, onFocus, onBlur]
}

export default useFocus
