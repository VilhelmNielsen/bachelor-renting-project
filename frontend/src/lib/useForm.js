import { useState } from 'react'

function useForm(initialValues) {
  const [form, setForm] = useState(initialValues)

  const onChange = event => {
    event.persist()
    const { name, value, type } = event.target
    setForm(prevForm => ({
      ...prevForm,
      [name]: type === 'checkbox' ? event.target.checked : value,
    }))
  }

  const reset = () => setForm(initialValues)

  return [form, onChange, reset]
}

export default useForm
