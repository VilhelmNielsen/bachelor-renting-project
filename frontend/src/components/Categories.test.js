import React, { Suspense } from 'react'
import { ApolloProvider } from 'react-apollo-hooks'
import { render, wait } from 'react-testing-library'
import { createClient, fakeCategory } from '../lib/testingUtils'
import Categories, { QUERY_CATEGORIES } from './Categories'

const category = fakeCategory()

const mocks = [
  {
    request: {
      query: QUERY_CATEGORIES,
    },
    result: {
      data: { categories: [category] },
    },
  },
]

describe('<Categories />', () => {
  fit('renders and matches snapshot', async () => {
    const client = createClient(mocks)
    const { getByText, container } = render(
      <ApolloProvider client={client}>
        <Suspense fallback="loading">
          <Categories />
        </Suspense>
      </ApolloProvider>
    )
    expect(getByText('loading')).toBeInTheDocument()
    await wait()
    expect(getByText(/Categories/i)).toBeInTheDocument()
    const link = container.querySelector('a')
    expect(link).toBeInTheDocument()
    expect(link.textContent).toBe(category.title)
    expect(link).toHaveAttribute('href', `/category/${category.slug}`)
    expect(container).toMatchSnapshot()
  })
})
