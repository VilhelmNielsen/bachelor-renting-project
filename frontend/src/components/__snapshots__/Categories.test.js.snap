// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Categories /> renders and matches snapshot 1`] = `
<div>
  <div
    class="css-anlc13-Layout"
  >
    <section>
      <h1>
        Categories
      </h1>
      <div
        class="css-1o89bq6-CategoryCard"
      >
        <a
          href="/category/tempora"
        >
          Quo quis molestiae
        </a>
      </div>
    </section>
  </div>
</div>
`;
